Source/Core/Core/NetPlayServer.cpp

```diff
@@ -2085,6 +2030,93 @@ bool NetPlayServer::PlayerHasControllerMapped(const PlayerId pid) const
          std::any_of(m_wiimote_map.begin(), m_wiimote_map.end(), mapping_matches_player_id);
 }
 
+void NetPlayServer::AssignNewUserAPad(const Client& player)
+{
+  for (PlayerId& mapping : m_pad_map)
+  {
+    if (mapping == Unmapped)
+    {
+      mapping = player.pid;
+      break;
+    }
+  }
+}
+
+PlayerId NetPlayServer::GiveFirstAvailableIDTo(ENetPeer* player)
+{
+  PlayerId pid = 1;
+  for (auto i = m_players.begin(); i != m_players.end(); ++i)
+  {
+    if (i->second.pid == pid)
+    {
+      pid++;
+      i = m_players.begin();
+    }
+  }
+  player->data = new PlayerId(pid);
+  return pid;
+}
+
+void NetPlayServer::SendResponse(MessageID message_id, const Client& player)
+{
+  sf::Packet response;
+  switch (message_id)
+  {
+  case MessageID::PlayerJoin:
+    response << MessageID::PlayerJoin;
+    response << player.pid << player.name << player.revision;
+    break;
+
+  case MessageID::ConnectionSuccessful:
+    response << MessageID::ConnectionSuccessful;
+    response << player.pid;
+    break;
+
+  case MessageID::ChangeGame:
+    response << MessageID::ChangeGame;
+    SendSyncIdentifier(response, m_selected_game_identifier);
+    response << m_selected_game_name;
+    break;
+
+  case MessageID::PadBuffer:
+    response << MessageID::PadBuffer;
+    response << m_target_buffer_size;
+    break;
+
+  case MessageID::HostInputAuthority:
+    response << MessageID::HostInputAuthority;
+    response << m_host_input_authority;
+    break;
+  default:
+    INFO_LOG_FMT(NETPLAY, "Warning! Call to SendResponse() failed to send a packet.");
+    return;
+    break;
+  }
```

Comment:
No... I don't like this.

The logic for what to send should remain in the main OnConnect flow, so it can be read inline. 

My goal with suggesting a SendResponse method was to try and separate the logic for building packets from the data that goes into them. Doing it like this just hoists random bits of logic from all over the place into this central location. 

Can I suggest using templates instead? 

I haven't tested this, but something like:

```c++
template<typename... Args>
void NetPlayServer::SendToPlayer(const Client& player, const MessageID message_id, Args&&... args)
{
  sf::Packet response;
  response << message_id;
  (response << ... << std::forward<Args>(args));
  Send(player.socket, response);
}
```

Then it can be used everywhere as just: 

```c++
   for (const auto& p : m_players)
  {
    SendToPlayer(player, MessageID::PlayerJoin, p.second.pid, p.second.name, p.second.revision);
    SendToPlayer(player, MessageID::GameStatus, p.second.pid, p.second.game_status);
  }
```

I'd then create a second version (probably an overload of `SendToClients`) for sending to all clients.

Reply:
Yeah, that makes sense. I've never used variadic arguments or variadic template arguments because they hurt my head. I'll read about them and see if I can figure out what you've written and put that in there.

Reply:
Confession: I had to google how to do the [folding](https://en.cppreference.com/w/cpp/language/fold) stuff. 

Advanced templates a very much a "sometimes" thing. They can massively help with cleaning up code; but they are very hard to understand.

Reply:
I think learning about folding added a few folds to my brain. It was an excellent suggestion though. I think the new code reads considerably better and is poised for some easy refactoring of other sections of NetPlayServer.cpp too. 

Unfortunately, I didn't see an easy way to tell the client about the selected game with the new SendResponse(), so I reverted that back to its old code.

Reply:
So, there is a solution:

We can refactor SendSyncIdentifier into an `sf::Packet << operator` overload

```c++
static sf::Packet& operator<< (sf::Packet& packet, const NetPlay::SyncIdentifier& sync_identifier) {
    packet << static_cast<sf::Uint64>(sync_identifier.dol_elf_size);

    packet << sync_identifier.game_id;
    packet << sync_identifier.revision;
    packet << sync_identifier.disc_number;
    packet << sync_identifier.is_datel;

    for (const u8& x : sync_identifier.sync_hash)
      packet << x;
    return packet;
}
```

(note, it has to be outside of the NetPlay namespace to work)

However, this arguably hampers readability. If having to jump to a function is considered hampering readability, then a custom operator overload is somewhat of a nightmare. The reader might not even notice its a custom overload, or know where to look. Some IDEs will allow to to ctrl+click on the overload to see where it goes, but others don't.

Reverting to the old code might just be the best solution. 

But... it could be argued that as long as it does the expected thing (send a data type down the wire), the reader doesn't need to know about it. We already define a few custom overloads in SFMLHelper.cpp

BTW, if you do want to go down this path, do it in another PR, and do the receive end too.